### ISSUE TITLE
[Req] cmd key support for Mac

### ISSUE THREAD

**Issue by pascallothar** [danny0838/firefox-scrapbook#108]
I have read in the manual of the legacy Scrapbook [http://www.xuldev.org/scrapbook/tips.php]:

- Click with Shift Key: Open in New Tab
- Middle-click: Open in New Tab
- Click with Ctrl Key: Open in New Tab

I am on _Mac Os X Snow Leopard 10.6.8 on a MacBookPro_, so I have no middle-click. Ctrl key behavior on Windows is usually operated by the Cmd key on Mac.

When I click on a item in the Scrapbook_X Sidebar with Cmd key down, it does the same than without the Cmd key, it opens the item in the current tab. This behaviour is not consistent with the usual behaviour on Mac and should open the item in a New Tab. And, so, I make often mistakes and open the item in a tab where I was doing some editing ... which goes lost.

Could you please implement the right behaviour for Mac or, at least, the possibility to choose it from the preferences?

**Comment by danny0838** [comment]
I have never used Mac. According to several documentation, it seems to that in the browser it is **Mac Control** key that acts like **Window Ctrl** key. Could you try Control+Click and see if it works?

**Comment by danny0838** [comment]
Additionally, could you test whether legacy ScrapBook work on Mac?

**Comment by pascallothar** [comment]
Hello,
I was wondering if I misled you, so I was googling a little bit to remember how it works on Windows and Linux.
Probably the documentations you mentionned were talking about the mapping of the keys **_themselves_** of the PC keyboards and the Mac keyboards (meaning hardware), but not about their functionality on the different operating systems.

For example, to **_copy_** on Windows, you will use `ctrl + C`, where you will use `cmd + C`on Mac.

[http://www.howtogeek.com/183636/how-the-command-and-option-keys-work-on-a-mac/]

On Mac,`Ctlr + Click` has everywhere the same behaviour than `right click` (probably because the earlier Macs had a mouse with only one button (meaning only the `"left" click` button)).

In [http://apple.stackexchange.com/questions/170731/how-to-open-link-in-a-new-tab-via-shortcut-in-chrome-on-mac]:

`Cmd⌘ + Click` _will open a link in a new tab behind the current one, if you click a link._
`Cmd⌘+ Shift⇧ + Click` _will open the new tab & bring it to the front._

**Those two behaviors** applied to a **Sidebar Item** would be consistent with the usual Mac behavior. For the moment, `Cmd + Click` doesn't do more than `Click`alone, and, as I said, is prone to "mistakes".

---

As you can see, Mac wouldn't have been Mac if they had done things easier and had followed the behavior of others :-)

**Comment by pascallothar** [comment]
I have just installed **_ScrapBook 1.5.13.1-signed.1-signed (Gomita)_** on my _MacBookPro with MacOsX 10.6.8 (last Snow Leopard)_.

Everything seems to work. Though I didn't test it thoroughly.

_**"Create MAF"**_ (your add-on version) on a **Sidebar Item** does _**NOT**_ work.
**_"Copy Page Info"_** (your add-on version) on a **Sidebar Item** **_DOES_** work, **_BUT_** does **_NOT_** work on the full tree (Sidebar -> **Tools -> Add-on functions** -> Copy Page Info: Arborescence complète).

---

Disabling _**Scrapbook**_, re-enabling **_Scrapbook X_** :-)

**Comment by danny0838** [comment]
Could you please check first whether control+click work on Mac? Likeweise, please also check whether control or cmd work on legacy ScrapBook?
If possible, check a few more ctrl usage cases mentionerd in the documentation you provided so that we can make sure wehether it's general case or not

**Comment by pascallothar** [comment]
Hello Danny,

1°) It seems that I didn't understand why you asked me to try the legacy Scrapbook on my Mac. I thought it was to help you to know if it was working or not on Mac, but, now, I have understood that it was to know its behavior with the Cmd and Ctrl keys. :-)

2°) I said: _« I was wondering if I misled you, so I was googling a little bit to remember how it works on Windows and Linux »_. I was not sure that I was remembering well the behavior on Windows and Linux, but I was sure of the behavior on Mac, since I use these shortcuts everyday :-) . (And also, _**I didn't change the default behavior**_). I remember it was a little bit annoying in the beginning when I migrated to Mac.

---
---

1°) The default behavior on Mac for `Ctrl+Click` (system wide) is to _open the contextual menu_ (Same behavior than if you have a mouse with a `right-click` (the earlier mouses branded by Mac had only one button)).

For _Firefox_, **_TabMixPlus_** allows you to change the behavior of `Ctrl+Click` from a°) _the default contextual menu_; to b°) _force opening in the current tab the stuff (depending on the other settings) that should open in an new tab._

There is no possibility to change the other behaviors of the key shortcuts like `Cmd+Click` or others from the Options set-up (perhaps is it possible with `about:config`?).

One of the two links I gave you [http://apple.stackexchange.com/questions/170731/how-to-open-link-in-a-new-tab-via-shortcut-in-chrome-on-mac] is for _Safari on Mac_ (the question was asked for _Chrome on Mac_, but the answer was mistakenly given for _Safari on Mac_). On _Safari_, you can change the default behaviour (but it will always use `Cmd` (actually, you will only toggle _new tab / new window_ in _new window / new tab_ and _background tab / active tab_ in _active tab / background tab_)), not on _Chrome_ (perhaps if you edit the config files?). In this link, you will see: « `Cmd + Alt + Click` = _Opens a link in a new window_ » and « `Cmd + Alt + Shift + Click` = _Opens a link in a new window, and makes it the active window_ », but it is SPECIFIC to _Safari_.
BUT « `Cmd + Click` = _Opens a link in a new tab_ » and « `Cmd + Shift + Click` = _Opens a link in a new tab, and makes it the active tab_ » are common to _Firefox_, _Chrome_, _Opera_ (I have never used _Opera_ before, but I have just installed it to check the behavior) and are not configurable and are the DEFAULT behavior of _Safari_. They sey also « _Remember that Cmd⌘ on Mac is usually equivalent to Ctrl^ on Windows_ »; I can say that, yes, it's USUALLY true.

Here is an other link [http://www.7is7.com/software/firefox/shortcuts.html] with the shortcuts for _Firefox 3.5_ and _3.6_ for the three major OS. You can see there (and I can confirm it) that on Mac OS X:
- `Cmd+Click` = _Open Link/Bookmark in new Background Tab_ (as said above, COMMON to all browsers).
- `Cmd+Shift+Click` = _Open Link/Bookmark in new Foreground Tab_ (as said above, COMMON to all browsers).
- `Shift+Click` = _Opens Link/Bookmark in New Window, and makes it the active window_ (Specific to _FIREFOX_).

---

2°) I will also make you the description of the behavior of the "Finder", of other file explorers or the like.
- `Click` = select.
- `Double-Click` = open.
- `Cmd+Click` = toggle select/deselect.
- `Shift+Click` = select all the items between the clicked item and the previously selected item.
- `Cmd+Double-Click` = open in a new tab (meaning for a single selected tab).
- `Cmd/Shift+Double-Click` = open in several new tabs (meaning for several selected tabs).

See my comment (#23 ).

---

3°) I have tried again **_ScrapBook 1.5.13.1-signed.1-signed (Gomita)_**:

a°) In the Sidebar (Gomita):

- `Click` = Open Item in the Current Tab :-) .
- `Double-Click` = Open Item in the Current Tab (nothing more than `simple Click`).
- `Shift+Click` = Select all the Items between the clicked Item and the previously selected Item AND open the clicked Item (only one) in a NEW Tab (How weird!!! An UNCLEAN mixed behavior!!).
- `Shift+Double-Click` = same, but will open the clicked Item twice (in two NEW Tabs).
- `Cmd+Click` = Toggle select/unselect the clicked Item (keeping the others selected) AND open the clicked Item in the CURRENT Tab (How weird!!!).
- `Cmd+Double-Click` = same than `Cmd+Click`.
- `Ctrl+Click` = Select the Item and open the contextual menu of the Item(s) (Doesn't open the Item in the tab; the contents of the tabs are not changed).
- `Right-Click` = same behavior than `Ctrl+Click`.
- `Ctrl+Double-Click` = nothing more than `Ctrl+Click`.
- `Alt+Click` = nothing more than `simple Click`.
- `Alt+Double-Click` = nothing more than `simple Click`.

b°) In the "Manage" window: (actually the behavior is the same than in Scrapbook X):

- `Click` = Select the Item.
- `Double-Click` = Select the Item and Open the Item in the CURRENT tab.
- `Cmd+Click` = Toggle select/deselect the Item.
- `Cmd+Double-Click` = Toggle select/deselect the Item AND Open the Clicked Item in the CURRENT tab.
- `Shift+Click` = select all the Items between (and including) the clicked item and the previously selected item.

	More precisely, if you click an item, it will be selected. Now `Shift+Click` a second item. All the items between the first and the second item will be selected. Now `Shift+Click` a third item. All the items between the FIRST (**_not the SECOND_**) and the THIRD will be selected (The mechanism acts as it will not remember an item as "beginning" when `Shift+Click`ed, but only when the last `simple Click`ed or the last `Cmd+Click`ed and will take for "end" the last `Shift+Click`ed. All the previously selected Items will be forgot).
- `Cmd+Shift+Click` = same, BUT doesn't forget the previously selected Items.
- `Shift+Double-Click` = Same than `Shift+Click`, but also open the Clicked Item (only one) in a NEW Tab.
- `Cmd+Shift+Double-Click` = Same than `Cmd+Shift+Click`, but also open the Clicked Item (only one) in a NEW Tab.
- `Ctrl+Click` = Open the contextual menu.

	More precisely (I am not sure if it is important for something), if you `Ctrl+Click` an item which is part of a bunch of already selected items, it will open the contextual menu and keep the selection as is. If you have a bunch of already selected items and you `Ctrl+Click` a not selected item, the previous selection will be lost, the contextual menu will be opened and the Clicked Item will be selected.
- `Ctrl+Double-Click` = nothing more than `Ctrl+Click`.
- `Right-Click` = exactly the same than `Ctrl+Click`.
- These combinations `+ Alt` = the same than these combinations `without Alt`.

---

4°) Now to compare with **_Scrapbook X (Version 1.13.0b4)_**:

a°) In the Scrapbook X Sidebar: (the differences with Gomita's are in bold italic):

- `Click` = Open Item in the Current Tab.
- `Double-Click` = Open Item in the Current Tab (nothing more than `simple Click`).
- **_`Shift+Click` = open the clicked Item in a NEW Tab._**
- **_`Shift+Double-Click` = open the clicked Item twice (in two NEW Tabs)._**
- **_`Cmd+Click` = same than `simple Click`._**
- **_`Cmd+Double-Click` = same than `simple Click`._**
- `Ctrl+Click` = Select the Item and open the contextual menu of the Item (Doesn't open the Item in the tab, the contents of the tabs are not changed).
- `Right-Click` = same behavior than `Ctrl+Click`.
- `Ctrl+Double-Click` = nothing more than `Ctrl+Click`.
- `Alt+Click` = nothing more than `simple Click`.
- `Alt+Double-Click` = nothing more than `simple Click`.
- `Cmd+Shift+Click` = same as `Shift+Click`.
- `Cmd+Shift+Double-Click` = same as `Shift+Double-Click`.

b°) In the "Manage" window:

Same behavior than in Gomita's Scrapbook.

---

5°) In the Firefox **_bookmarks_** window:

- For the **_combinations with `simple Click`_**, same behaviour than in the "Manage" window.
- `Double-Click` = Select the Item and Open the Item in the CURRENT tab.
- If no item is already selected, `Cmd+Double-Click`ing an item will open it in a NEW TAB.

	Otherwise nothing more than for `Cmd+Click` will happen.
- If only one item is already selected, `Shift+Double-Click`ing it will open it in a NEW WINDOW.

	Otherwise nothing more than for `Shift+Click` will happen.
- The **_other combinations with `Double-Click`_** will do the same than with `simple Click`.

---

6°) I wonder if, this night, I will dream of:

- `Simple Click`s?
- `Double Click`s?
- `Cmd+Click`s?
- `Shift+Click`s?

Or ... of all of them mixed together? :-)
Sure, it will be the ... right click! :-)

**Comment by danny0838** [comment]
In both references on the [Mozilla Developer Network](https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/metaKey) and [W3school](http://www.w3schools.com/jsref/event_metakey.asp), it is a standard that a mouse event with metaKey = Win key on Window = Command key on Mac. There is no specific documentation talking about ctrlKey, but I think it is Control key on Mac since Command key is used by metaKey.

Maybe it's custom that Mac users use command as Ctrl in Windows, but it already a standard that in the browser javascript system it's Command = metaKey = Win key. We don't know why the standard is formulated in this way, but we have to add many additional codes and ac hoc handlers to do the remap for Mac to work use Command key as Ctrl.

Here are still something we need to know:

1. Is Control key on Mac naturally work as a key for context menu?

2. Could you try the hotkeys for DOM erasor and for HTML editor and see whether Command or Ctrl key work? 
(Open a captured page, and you can see DOM erasor and HTML editor buttons in the editor toolbar. Hotkeys for DOM erasor are listed when you activate the DOM erasor. Hotkeys for HTML editor is shown in the context menu when HTML editor is activated )

**Comment by danny0838** [comment]
@pascallothar On 1.13.0b8 we implemented a new hotkey system to support cross-platform issues. Welcome to try whether `command` key work as expected and whether hotkeys are showed in Mac style.

This is the hotkeys shown on most platform:
![screenclip](https://cloud.githubusercontent.com/assets/531417/17034914/4ef78ed2-4fb7-11e6-912e-8bfe49ce0dbc.png)

This is the hotkeys ought to shown on Mac if it works right:
![screenclip 1](https://cloud.githubusercontent.com/assets/531417/17034916/4fda1d4c-4fb7-11e6-8de3-87e0aa40ac42.png)

**Comment by pascallothar** [comment]
The answers are for your TWO posts:

---
---

> In both references on the [Mozilla Developer Network] and [W3school], it is a standard that a mouse event with metaKey = Win key on Window = Command key on Mac. There is no specific documentation talking about ctrlKey, but I think it is Control key on Mac since Command key is used by metaKey.
>
> Maybe it's custom that Mac users use command as Ctrl in Windows, but it already a standard that in the browser javascript system it's Command = metaKey = Win key.

Yes, you are right:

- `event.ctrlKey` ==> `Ctrl` key (also noted caret `^` ) on Windows, Linux and Mac. Actually, nothing illogical here. Did Mac (actually Windows (see history)) forget to show its difference, here? :-)
- `event.altKey` ==> `Alt` key on Windows, Linux and Mac ( `⎇` ) (on Mac the key is also named `Option` and labelled `⌥` ).
- `event.shiftKey` ==> `Shift` key (also labelled `⇧` ) on Windows, Linux and Mac (no bad luck there, fortunately :-) .

 Those keys are historically quite older (to be precise, they were more widespread among different systems) than the following one (to be precise, the following key was already existing before it was "re-introduced" by Windows and Mac):
 
- `event.metaKey` ==> On Windows: `Windows` key ( `⊞` ).
                      ==> On some "agnostic" keyboards, it is labelled `Meta` or `◆` to please, for example, some Linux and BSD "fanatic" users.
                      ==> On Mac keyboards, it used to be labelled with an apple icon ( `` ); nowadays, it is labelled `Command`, `⌘` or `Cmd`.
                      Actually, if you plug a Windows branded keyboard or a generic keyboard to a Mac, it works immediately and you have to use the `Windows` key where you would have used the `Cmd` key with a Mac keyboard.
                      NOTES: **_1°)_** On some "older" systems laking a `Meta` key, there were work-arounds using `Alt` (or even `Space` as a modifier). **_2°)_** Sometimes, `Meta` key is named `Super` key (although, on other systems like Lisp, `Meta` and `Super` are different keys), but, in spite of that, it doesn't fire a `event.superKey` event, but the `event.metaKey` event. Yes, in computer sciences, the things you could think to be the simplest things are everything but simple :-) .

---
 
> Maybe it's custom that Mac users use command as Ctrl in Windows, but it already a standard that in the browser javascript system it's Command = metaKey = Win key. We don't know why the standard is formulated in this way, but we have to add many additional codes and ac hoc handlers to do the remap for Mac to work use Command key as Ctrl.

        Yes, the lack of coordination and of standardization is a nightmare for informatics (and for the world in general (kind of Babel tower)).
        As I understand things from the stuff I have read in the past (and today), `Shift` was already used by typewriters; `Ctrl` (as its name stands for), `Alt`, and other control keys and control characters were used by the teletypewriters, the terminals, the terminal emulators and the earlier keyboards of different electro-mechanical, electrical or computer systems. Xerox begun the GUI operating system, followed by the Apple Macintosh, followed by Microsoft Windows. They reused features of the legacy systems, but they were following their own way of doing things. The portability of the applications was not their preoccupation at that time. As for the war of document and media formats that happened at those times, the war between Mac and Windows was not to ease the standardization, which occurred much later because of the needs of industry. Fortunately, nowadays, a lot of the most common shortcuts have perhaps a different modifieR key, but the modifieD key is the same. Imagine if Mac would use Cmd+Q, Linux Ctrl+D and Windows Ctrl+C to copy. We could of course imagine worse: that the British oranges would be American bananas and that British bananas would be American oranges. Or that the words "on" and "off" would have inverted signification in the two countries. Yes, it's almost that with the `Cmd` and `Ctrl` behaviors :-) .

---

> Is Control key on Mac naturally work as a key for context menu?

        No. `Ctrl` on Mac, as on Windows, is a modifier key, meaning that if you push it alone, nothing happens (USUALLY, to be precise! Because some rare applications will use a modifier key as a simple key (hitting once or twice `Alt` (or and other modifier) or keeping it pushed more than 2 seconds for example), but I don't remember any example for the moment. It is also possible to use a simple key as a modifier (for example, when `Space` is down, hitting K). Actually, this is not specific to Mac).
        But `Ctrl+Click` is the equivalent of `Right-Click` (As I said above and in other posts, Mac used GUI and mouse with only one click before MicroSoft created Windows, so, the contextual menu was triggered by `Ctrl+Click`. Only later, Mac introduced a "Secondary Click" (name of the right-click in the Mac world, probably to not use the same word than Microsoft)).
        Besides `Cmd` and `Shift`, `Ctrl` and `Alt` are used for shortcuts also, but they are not so many than the shortcuts build with `Cmd` == `Meta`. [Mac keyboard shortcuts https://support.apple.com/en-us/HT201236] will give you the shortcuts used system-wide and also the shortcuts used by the applications branded by Mac and integrated in their distribution. The third party applications will try to mimic the behavior of the Mac "integrated" applications (e.g.: Firefox shortcuts are usually the same than Safari's (Mac "integrated" web browser)).

---

Here are some links concerning what I said above in the present post:

- [https://en.wikipedia.org/wiki/Keyboard_layout#Modifier_keys] especially §5 and §7.
- [https://en.wikipedia.org/wiki/Keyboard_layout#History] especially §7 and §9.
-   
- [https://en.wikipedia.org/wiki/Control_key#History]
- [https://en.wikipedia.org/wiki/Control_key#Similar_concepts]
-   
- [https://en.wikipedia.org/wiki/Modifier_key#Modifier_keys_on_personal_computers] §2
-   
- [https://support.apple.com/en-us/HT201236]

---

> Could you help me check an Firefox config on Mac? Type about:config in the address bar, and find the value of ui.key.accelKey.

ui.key.accelKey == 224

---

In the "Version History":
> Functionality changes:
>
>  - Improve cross platform hotkey support, especially the command key support and the key symbol style for Macs.
>
>  - Allow customized hotkey combination in the user options.
>

I don't see that in the Options GUI? Is it in a configuration file, then?

---

> .../... and whether hotkeys are showed in Mac style.
> .../...
> This is the hotkeys ought to shown on Mac if it works right:

No. Here is what Mac users ACTUALLY see:

![html_editor_mode_contextual menu](https://cloud.githubusercontent.com/assets/20140037/17193125/1b7874d4-5441-11e6-980b-34a81bab67c4.png)


For me and aware people :-) , it's not a problem, but very probably, most people would have no clue of what `Meta` is!

---

> Could you try the hotkeys for DOM erasor and for HTML editor and see whether Command or Ctrl key work?

# First is the test with 1.13.0b7:

### DOM Eraser:
- NOTE: There is no use of `Ctrl` or `Cmd` in DOM Eraser itself.
- All the hot-keys work.
- Because `F9` is used BY DEFAULT system-wide to display on the Desktop the thumbnails of all the windows open or minimized (myself, I use 4-fingers gesture on the trackpad), you need to:
_"System Preferences ..."_ -> _"Keyboard"_ -> _"Keyboard Shortcuts"_ -> _"All windows (F9)"_ -> uncheck the checkbox or change the default value.
- Following is a description of some behaviors that can perhaps help you to understand how the "signal" is passed by Firefox to _Scrapbook X DOM Eraser_ and then, if not "stopped" ("intercepted" and "deleted" vs "intercepted" and "passed"), by _Scrapbook X DOM Eraser_ back to Firefox.
- `Right-Click` will _**isolate**_, but will also open the "_contextual menu_". It's not a big problem, since it is easy to close the "_contextual menu_" and there are also other hot-keys to _**isolate**_. Is it the same on Windows?
- If I hit a non-hot-key, for example `x` , _"Find in the page ..."_ bar will open with `x` written in it. It's also the default behaviour when not in _Scrapbook X DOM Eraser_ mode.
Now, if some text were already selected before entering in _Scrapbook X DOM Eraser_ mode, hitting `w` will _**widen**_ as expected, but almost as soon, a new tab will be opened with the selected text searched on Wikipedia, which is the behavior when not in _Scrapbook X DOM Eraser_ mode (because I have installed an add-on doing this).
- On my MacBookPro with a FRENCH keyboard layout, if I want to write a `+` , I need to hit `shift`+`=` (`=` with `shift` down) or `Caps_lock`+`=` .
In _Scrapbook X DOM Eraser_ mode, surprisingly, `shift`+`=` will work as `w` does. Probably because there is no hot-key defined for `shift`+`=` .
Surprisingly, `Caps_lock`+`=` will not work. It will open the _"Find in the page ..."_ bar with `=` written in it. 
- On a captured page, I didn't notice anything else. But with "_**Edit before capturing**_", on a GitHub page, when I use `c` to _**colorize**_, it is **_colorizing_**, but soon after this, the page will change. I have understood that hitting `c` in a GitHub page will open a page for a new issue. But all the annotations will be lost. It's better to be known.
- Are those behaviors normal? Or should Scrapbook X catch the event when in _DOM Eraser mode_ and not pass it back to Firefox? Is it specific to Mac or do you have the same behavior on Windows?

### HTML Editor:
- I have tested all the hot-keys in the "main" sub-menu and in the sub-SUB-menus (Text, Block, Insert, Advanced). Most of them work. Following is what I noticed.
- _**Undo**_ inaccessible when in _Scrapbook X HTML Editor_ mode. Would it be possible to make it accessible or to implement a hot-key for it?
- Because `F10` is used BY DEFAULT system-wide to display on the Desktop the thumbnails of all the windows open or minimized OF the current APPLICATION, you need to:
_"System Preferences ..."_ -> _"Keyboard"_ -> _"Keyboard Shortcuts"_ -> _"All application windows (F10)"_ -> uncheck the checkbox or change the default value.
- (EDIT: See below in 1.13.0b9 test before). `Ctrl+M` doesn't do anything. With the Firefox add-on [Menu Wizard https://addons.mozilla.org/fr/firefox/addon/s3menu-wizard/], I can see that `Ctrl+M` is used by `key_toggleMute`, but there is no menu entry for it!!! Very weird! I unchecked this hot-key to disable it. Even after restarting Firefox, `Ctrl+M` doesn't do anything in _Scrapbook X HTML Editor_ mode or not!!! But I am able to **_Clear Format_** from the contextual menu. `Ctrl+M` isn't used by the system neither. I wonder if `key_toggleMute` is not some unused relic that was used to toggle the audio Firefox-wide before the implementation of loudspeaker icons on every tab to toggle mute/unmute the audio output of the tab. So, perhaps is `Ctrl+M` caught by this relic and "lost in the vacuum"? (EDIT: Very seldom, it works!!!!! So I tried an external American layout keyboard, but it's the same unstable behavior).
- To write `[` and `]`, I need to hit `Alt+Shift+(` and `Alt+Shift+)` , so I am unable to `Alt+[` and `Alt+]` (to **_Outdent_** or **_Indent_**). With the American keyboard, of course, it works fine.
- To write `.` and `/`, I need to hit `Shift+;` and `Shift+:` , so I am unable to `Alt+.` and `Alt+/` (to **_Align Right_** or **_Justify_**). With the American keyboard, of course, it works fine.
- **_Align Center_** from the contextual menu works fine, but when I hit `Alt+M`, the behavior is different, not good (Some change in the layout of the text, but not the expected one). As expected, same bad behavior with the American keyboard.

# Now is the test with 1.13.0b9:

In the Options dialog, "Show the Sidebar" == `Alt+K` works.
The newly implemented "Menu" == `Alt+C` does nothing. Which "Menu" is it?

### DOM Eraser:
- Same behavior as with 1.13.0b7, except for `-` which will not work anymore but only open the _"Find in the page ..."_ bar with `-` written in it.
- And `Caps_lock`+`=` will now open the _"Find in the page ..."_ bar with `+` written in it (with 1.13.0b7, it was `=` !!!).

### HTML Editor:
- Almost the same behavior as with 1.13.0b7. Following are the differences that I noticed:
- A lot of combinations with the `Cmd` key are already used by Firefox. When one of these combinations is successfully passed by Firefox to _Scrapbook X HTML Editor_, the menu bar will blink once (e.g.: Firefox use `Cmd+B` to "View" -> "Sidebar" -> "Bookmarks". When you hit `Cmd+B` while in _Scrapbook X HTML Editor_ mode, "View" in the Menu bar will blink once in blue and then the selected text in _Scrapbook X HTML Editor_ will be styled in bold). Fewer other combinations are also used by the system or system-wide, but will not be passed (I am not sure if it is for all of them) to an application, so not to Firefox and, a fortiori, not to _Scrapbook X HTML Editor_.
- `Cmd+Shift+C` is used system-wide by Clyppan, a third party application that is an extended clipboard app. While in _Scrapbook X HTML Editor_ mode, Clyppan will work as usually and nothing will happen in _Scrapbook X HTML Editor_.
- `Cmd+Alt+8` is used by the system to handle the zoom for disabled people. While in _Scrapbook X HTML Editor_ mode, hitting `Cmd+Alt+8` will work as usually and nothing will happen in _Scrapbook X HTML Editor_. I don't use this feature, so I go to "System Preferences" to disable it. Now it works in _Scrapbook X HTML Editor_ as expected.
- The hot-key `Cmd+M` for **_Clear Format_** has the same bad behavior than `Ctrl+M` with 1.13.0b7. If Firefox would have not passed `Cmd+M` to _Scrapbook X HTML Editor_, Firefox would have kept it for itself and minimized the window, but it didn't. So I can conclude that this bad behavior has nothing to do with a certain relic (as I thought), but that _Scrapbook X HTML Editor_ has caught `Cmd+M` (or `Ctrl+M` in 1.13.0b7), but didn't handle it properly to trigger **_Clear Format_**.

---

As you can see, you had a good idea in allowing customized hotkey combination in the user options.

---

NOTE: I didn't say that it is bad to use `Ctrl` on a Mac, but that:
- `Ctrl+LeftClick` doesn't work the way you thought (see above).
- `Ctrl+"key"` works, but is often not consistent with the shortcut of an other application using the same or a similar function and that would use `Cmd+"key"`. For app or add-on specific functions, there is no problem to use `Ctrl` or `Alt` or `Shift` or `Cmd` or even all of them together to implement a shortcut. Actually, they are often used by Firefox add-ons when it comes to implement novel features. And they will have less chance to conflict with other shortcuts, since the canonic shortcuts very often use `Cmd`.

---

Would it be difficult to implement some Help pop-up in _Scrapbook X HTML Editor_ mode (I would say with shortcut `Cmd+?`) like the one in _Scrapbook X DOM Eraser_ mode?

---

I have looked up the usage of shortcuts in different HTML editors on the Mac OS X platform (iWeb, Komposer, BlueGriffon, Adobe Brackets, Adobe Dreamweaver, SeaMonkey Composer, Amaya). There is no real standard usage, only some canonical common shortcuts and some others poorly shared (As you can see once more, you had a good idea in allowing customized hotkey combination in the user options). When I will be able to access the "customized hotkey combination in the user options", I think I will choose (But I don't say it is the best or the only way to make a default usage on Mac. I believe that people would choose according to the app they are used to work with) the following customization (so, I give it only for simple info):

- `F10` == "switch html editor" _== Same as SB Windows version_
- `^Q` == "Quit html editor" _== Not implemented_ ( `⌘Q` could quit Firefox)
- `⌘?` (actually `⌘,` because of my keyboard layout) == "Help" _== Not implemented_
- `⌘S` == "Save the page" _== Same as CURRENT SB Mac version_
- `⌘Z` == "undo" _== Not implemented_
- `⇧⌘Z` == "redo" _== Not implemented_
- `^⌘Z` == "clear undo history" _== If implemented (if useful?)_
- `^N` == "clear liNks" _== Same as SB Windows version_
- `^M` == "clear forMat" _== Same as SB Windows version_
- `⌥⌘I` == "Insert or edit source" _== Same as CURRENT SB Mac version_
- `⌘B` == "switch Bold" _== Same as CURRENT SB Mac version_
- `⌘I` == "switch Italic" _== Same as CURRENT SB Mac version_
- `⌘U` == "switch Underlined" _== Same as CURRENT SB Mac version_
- `⌘K` == "switch strucK-through"
- `⌘↑` == "switch superscript"
- `⌘↓` == "switch subscript"
- `⌘+` (actually `⌘=` because of my keyboard layout) == "increase font size"
- `⌘0` == "restore font size" _== Not implemented_
- `⌘-` == "decrease font size"
- `⌘E` == "set tExt and background color..." _== Same as CURRENT SB Mac version_
- `⌥[` (actually `⌥(` because of my keyboard layout) == "increase indent"
- `⌥]` (actually `⌥)` because of my keyboard layout) == "decrease indent"
- `⌥←` or `⌥L` == "align Left"
- `⌥→` or `⌥R` == "align Right"
- `⌥↑` or `⌥M` == "align center (Middle)"
- `⌥↓` or `⌥J` == "Justify"
-   
- And for the other shortcuts, I would use the _same as SB Windows version_.

---

I hope it will help you.
Pascal

**Comment by danny0838** [comment]
@pascallothar We need your help to get it work right.

Please download and install the .xpi file in the [test version](https://github.com/danny0838/firefox-scrapbook/releases/tag/1.13.0b9.1-osdetector) (you probably have to **set xpinstall.signatures.required to false in the "about:config" to make it installable** on your Firefox). If it work correctly, an alert window will prompt and report your OS when you restart your Firefox like this:

![sp160730_222534](https://cloud.githubusercontent.com/assets/531417/17271109/fde712ec-56a5-11e6-9ba0-3fc26017336d.png)

Please report the text content or the screenshot image. We need this information to fix the OS detection issue on Mac.

**Comment by pascallothar** [comment]
I have only one Mac:
It's a laptop:
MacBookPro
with Mac OS X 10.6 (also named "Snow Leopard"), 10.6.8 precisely.

![for_danny](https://cloud.githubusercontent.com/assets/20140037/17273733/824c47e2-56af-11e6-8b9b-62e114b33708.png)

**_xpinstall.signatures.required_** to **_false_** was already needed to install 1.13.0b7 (or b5 or something, I don't remember)

Question: I suppose that you **_DID_** know that Scrapbook X **_DID_** already detect Mac (because it was using Cmd key in the Manage window). Didn't you? I prefer to tell you it, to avoid you extra-work.

**Comment by danny0838** [comment]
Here is the [second test version](https://github.com/danny0838/firefox-scrapbook/releases/tag/1.13.0b9.2-shortcuttest). Please check:

1. Whether key names are shown in Mac style on Macs.

2. Whether the default behavior of shortcut keys are still not skipped. (e.g. right-click isolate for DOM Eraser shows a context menu, )

3. Whether customization of HTML editor shortcut keys via the options dialog work right.

Notes:

1. When HTML Editor mode is on, `Ctrl+A` (Select all), `Ctrl+C` (Copy), `Ctrl+X` (Cut), `Ctrl+V` (Paste), `Ctrl+Z` (Undo), `Ctrl+Shift+Z` or `Ctrl+Y` (Redo), etc, are already supported natively by Firefox. Please check whether they actually work or not.

2. We probably cannot solve the issue that system wide hotkeys overwrites our hotkey since systemic hotkey takes precedence. The only way is to modify the system hotkey or to modify our hotkey to prevent a conflict.

**Comment by pascallothar** [comment]
> Whether key names are shown in Mac style on Macs.

Yes, now, they are.
Actually, the `+` sign between the keys are usually not print, but IMHO, it doesn't matter.


- _Example of Mac menu:_
   
![cmd z greyed out](https://cloud.githubusercontent.com/assets/20140037/17280829/2442abe2-5785-11e6-834c-ae69cccf82f5.png)


- _Main HTML Editor menu:_

![a main](https://cloud.githubusercontent.com/assets/20140037/17280835/45cbc672-5785-11e6-9ee1-84a02649301c.png)

- _--> "Text":_

![b text](https://cloud.githubusercontent.com/assets/20140037/17280864/c8dab866-5785-11e6-9f0a-893e10b2cea1.png)

- _--> "Block":_

![c block](https://cloud.githubusercontent.com/assets/20140037/17280870/dd55de88-5785-11e6-8a44-c4cb67232b77.png)

- _--> "Insert":_

![d insert](https://cloud.githubusercontent.com/assets/20140037/17280876/efd84456-5785-11e6-9502-e9afde820254.png)

- _--> "Advanced":_

![e advanced](https://cloud.githubusercontent.com/assets/20140037/17280886/0393581e-5786-11e6-8c33-8d5c8823d5fd.png)

---
---

> Whether the default behavior of shortcut keys are still not skipped. (e.g. right-click isolate for DOM Eraser shows a context menu)

Actually, except for (1) and (2) below, everything is OK.

## DOM Eraser

- `w` works now perfectly in 1.13.0b9.2-shortcuttest.
- `c` works now perfectly in 1.13.0b9.2-shortcuttest.
- As said in the post above, `-` was working perfectly in version 1.13.0b7.
  - But in version 1.13.0b9, it was working as a normal key (not anymore as a hot-key), meaning that it will open the "_Find in the page ..._" bar with `-` written in it.
  - Now, in 1.13.0b9.2-shortcuttest, it works the same (wrong) way as it was already working in 1.13.0b9. (1)

Nothing else have changed since 1.13.0b9 (see in the post above for the detailed feedback).
So, in 1.13.0b9.2-shortcuttest:

- `Right-Click` will _**isolate**_, but will also open the "contextual menu". (2)
- The behavior with `Caps_lock` is IMHO not a real issue, since the hot-keys are now customizable.

## HTML Editor

- The hot-key `Cmd+M` for **_Clear Format_** now works perfectly in 1.13.0b9.2-shortcuttest.
- The hot-key `Alt+M` for **_Align Center_** now works perfectly in 1.13.0b9.2-shortcuttest.

All the others things were already working the way expected to be.

---
---

> Whether customization of HTML editor shortcut keys via the options dialog works right.

The default size of the Options Dialog is different than it was. It's totally collapsed for several tabs, but, because I was able to resize it, it is OK. But, for the tab of the shortcuts, I couldn't access all the shortcuts, because the dialog would need more height than my screen. So I would need to find on my disk the Scrapbook X configuration file and edit it by hand. But I didn't find it. Here are screenshots of the Options Dialog.

- 1st tab as it opened the first time:
   
   ![1 general](https://cloud.githubusercontent.com/assets/20140037/17280996/97676476-5787-11e6-948e-057248aa99b4.png)
   
   -  1st tab resized:
       
       ![1 general resized](https://cloud.githubusercontent.com/assets/20140037/17281003/a8d1ae42-5787-11e6-98be-a76b628229a9.png)

- 2nd tab was already at maximum height, which is NOT ok:
   
   ![2 keys was already at maximum size not ok](https://cloud.githubusercontent.com/assets/20140037/17281007/b5434f8c-5787-11e6-8fb9-e36aed8a90f7.png)
   
- 3rd tab:
   
   ![3 tabs](https://cloud.githubusercontent.com/assets/20140037/17281043/36910642-5788-11e6-865b-1bd427973a2c.png)
   
   - 3rd tab resized:
      
      ![3 tabs resized](https://cloud.githubusercontent.com/assets/20140037/17281045/42169fea-5788-11e6-872c-856255380d76.png)
      
- 4th tab at maximum size is OK:
   
   ![4 edit at max size ok](https://cloud.githubusercontent.com/assets/20140037/17281049/4cedcf9c-5788-11e6-8a6a-fb8834cb8456.png)
   
- 5th tab: As you can see, even if resized very high, it will not show everything without a scroll bar.
   
   - Top:
   
      ![5 organize resized top](https://cloud.githubusercontent.com/assets/20140037/17281057/68f8da38-5788-11e6-9ad3-5f1e16396ff6.png)
      
   - Bottom:
   
      ![5 organize resized bottom](https://cloud.githubusercontent.com/assets/20140037/17281064/77e31c34-5788-11e6-9408-3b23710353ca.png)
      
- 6th tab: Same comment as for 5th, but you can see also the horizontal scrollbar. Perhaps it can help you to see how to fix the problem of the 2nd tab, which has no scrollbar, but perhaps would need one.
   
   - Top:
      
      ![6 advanced resized top](https://cloud.githubusercontent.com/assets/20140037/17281071/820e2938-5788-11e6-8a96-68b2e9ab217c.png)
      
   - Bottom:
         
      ![6 advanced resized bottom](https://cloud.githubusercontent.com/assets/20140037/17281074/8cf390ae-5788-11e6-9cc0-d2c83e979a27.png)


I have edited the hot-keys in the 2nd tab. They don't update in the contextual menu nor do they update in their behavior. So I restarted Firefox. _**Now it works.**_

For the inaccessible hot-keys and for those of DOM Eraser, I have tried to find a config file, but I could not find it!

---
---

> When HTML Editor mode is on, Ctrl+A (Select all), Ctrl+C (Copy), Ctrl+X (Cut), Ctrl+V (Paste), Ctrl+Z (Undo), Ctrl+Shift+Z or Ctrl+Y (Redo), etc, are already provided natively by Firefox. Please check whether they actually work or not.

Yes, they work (Cmd for Mac of course).

---
---

> We probably cannot solve the issue that system wide hotkeys overwrites our hotkey since systemic hotkey takes precedence. The only way is to modify the system hotkey or to modify our hotkey to prevent a conflict.

Actually, it is not even desirable that system wide hot-keys would be overridden. I was only making a complete feedback of the behavior for a better understanding. And to report that some hot-keys chosen by Scrapbook were already taken by the system. And since the hot-keys are now customizable, it should not be seen as an issue anymore.

---
---

On the page of the current version, I see:

> Customizable DOM Erasor shortcuts. (Only via user perference)

I cannot find out how to customize them. What and where is this "user preference"?

---
---

The newly implemented "Menu" == Alt+C does nothing. Which "Menu" is it?

Actually, in "HTML Editor" mode, it writes only `©`, which is the default behavior if there is no shortcut defined.

**Comment by danny0838** [comment]
> I cannot find out how to customize them. What and where is this "user preference"?

It means Firefox user preference, which are all available in `about:config`, and some of them are available in the GUI options.

Note that the user preference uses the normalized string which could be different from the GUI. If you don't know how to write, you can set a few key combinations (e.g. `Show in Sidebar`) in the GUI and see how corresponding user preferences (e.g. `extensions.scrapbook.key.sidebar`) changes in the `about:config`.

We do not provide the GUI way because we do not encourage a modification on DOM Erasor keys.

> Actually, the `+` sign between the keys are usually not print, but IMHO, it doesn't matter.

We'll adjust the accesskey text to be more compatible to Mac style.

> Right-Click will isolate, but will also open the "contextual menu". (2)

Please check whether `dom.event.contextmenu.enabled` is set to `false` in the `about:config` in your Firefox. If yes, set it to `true` and try again.

I'm not sure whether `Control+Click` is also affected by this setting. You can also test it for Control+Click issues.

> - As said in the post above, `-` was working perfectly in version 1.13.0b7.
>   - But in version 1.13.0b9, it was working as a normal key (not anymore as a hot-key), meaning that it will open the "_Find in the page ..._" bar with `-` written in it.
>   - Now, in 1.13.0b9.2-shortcuttest, it works the same (wrong) way as it was already working in 1.13.0b9. (1)

The accesskey system only detects whether the configured key and modifier keys (i.e. Ctrl, Alt, Shift, etc) are pressed. Which means that:
- It does not change by holding `Shift`. If your keyboard has key `(` that outputs `[` with `Shift` hold, it is simply a `(` key or `Shift+(`, which is never equivalent to `[` for the accesskey system.
- It does not change by active `CapsLock`. If your keyboard has a key `(` that outputs `[` with `CapsLock` active, it's simply a `(` key for the accesskey system.

When the DOM Eraser says `+` and `-`, it actually says `Shift+=` or `Numpad+`, and `-` (hyphen-minus) or `Numpad-`. Please confirm what keys are actually pressed when you mean `-` on your keyboard. Is there a modifier key?

> But, for the tab of the shortcuts, I couldn't access all the shortcuts, because the dialog would need more height than my screen. So I would need to find on my disk the Scrapbook X configuration file and edit it by hand. But I didn't find it. Here are screenshots of the Options Dialog.

I really have no idea why it works so on your computer. All option tabs are written in the same way and there should be scrolls for the hotkey tab just like other tabs do if it works normally:

![dialog](https://cloud.githubusercontent.com/assets/531417/17283830/7e14a7bc-57e7-11e6-9471-fe96786a0abc.jpg)

A guess is that it will work so if the total contents overflows your screen. But it ought not work like this, and it it does, it should be a bug of Firefox.

Could you check whether resizing the dialog window works? And whether mouse scroll has an effect?

Also, hotkeys for HTML Editor are also user preferences and are available in the `about:config`. You can set them there is the GUI is unavailable.

**Comment by danny0838** [comment]
Here is the [third test version](https://github.com/danny0838/firefox-scrapbook/releases/tag/1.13.0b9.3-macpref). Please check:

1. Whether the Mac style hotkey text work right.

2. Whether the scrollbar works in the options panel if the content overflows the dialog window.

**Comment by pascallothar** [comment]
> Please check `about:config` and see whether `dom.event.contextmenu.enabled` is set to `false`. This will block any attempt to block the context menu from showing up so a right click always shows the context menu.

`dom.event.contextmenu.enabled` == `true`
If I have well understood, it is the right value. Isn't it?

> I'm not sure whether `Control+Click` is also affected by this setting. You can also test it for Control+Click issues.

Could you be more explicit?

> > On my MacBookPro with a FRENCH keyboard layout, if I want to write a + , I need to hit Shift+= (= with Shift down) or Caps_lock+= ...

> If your keyboard has key = that outputs + with Shift hold, it is simply = or Shift+=, which is never equivalent to + for the accesskey system.

> Similarly, if your keyboard has a key = that outputs + when CapsLock is active, it's simply = for the accesskey system.

> But we are somewhat lazy. When the DOM Eraser says + and -, it actually means Numpad+ or Shift+= and Numpad- or - (hyphen-minus).

Thank you. It helps me to better understand how the French keyboard layout works.

Indeed, by playing with the customization fields, I can see that, on my keyboard, even if I have to hit `(` with `Shift` down or `(` with `Caps_Lock` to write a `5`, the customization field will be filled with `5` when I hit `(` with or WITHOUT `Caps_Lock`.

Meaning that I will never be able to define a shortcut `Alt+(` (to replace `Alt+[` which is not accessible with my keyboard), but only `Alt+5` (that is already used).

But `Alt+)` is possible (`)` with `Shift` down writes a `°`).

Meaning that the mapping of the French keyboard layout to the "access-key system" use sometimes "uppercase", sometimes "lowercase", and that independently of `Caps_Lock`.

> > - As said in the post above, `-` was working perfectly in version 1.13.0b7.
> >   - But in version 1.13.0b9, it was working as a normal key (not anymore as a hot-key), meaning that it will open the "_Find in the page ..._" bar with `-` written in it.
> >   - Now, in 1.13.0b9.2-shortcuttest, it works the same (wrong) way as it was already working in 1.13.0b9. (1)
> > 
> 
> Please check what keys are actually pressed when you mean - on your keyboard. Is there a modifier?

No, no! No modifier. It is the `hyphen-minus`.

`extensions.scrapbook.key.domEraser.narrower3` was set to `Hyphen_Minus`. I have deleted the `_`, setting it to `HyphenMinus`. Now it works again (as it did in 1.13.0b7).

Are you sure that it works on your Windows with the string `Hyphen_Minus` (with `_`)? If it does, Steve Jobs (or somebody else, probably) needs some posthume ears pulling :-) to learn not being different uselessly.

> And please also check the about:config and see whether accessibility.typeaheadfind is set to true. This determines whether a key fires a search, and it seems to block certain accesskeys from working in some unclear situation.

`accessibility.typeaheadfind` == `true`

Actually the key fires a search only if it is not defined as a hot-key, so it should not be a problem. As said above, `-` was not defined because the string `Hyphen_Minus` was not recognized as `-`.

> Here is the third test version. Please check:
>
>  1. Whether the Mac style hotkey text work right.
    
Yes, it works right. Here are the screen-shots.
  
  
![menu1](https://cloud.githubusercontent.com/assets/20140037/17388913/1af37ff0-59ef-11e6-955c-16cf4e174506.png)

![menu4](https://cloud.githubusercontent.com/assets/20140037/17388936/4262251e-59ef-11e6-8376-e4345cdeabfe.png)

![menu3](https://cloud.githubusercontent.com/assets/20140037/17388939/47f75166-59ef-11e6-97c6-785727b4f2f8.png)

![menu2](https://cloud.githubusercontent.com/assets/20140037/17388943/4d9a3fc0-59ef-11e6-8bed-5c58f43f1949.png)


>  2. Whether the scrollbar works in the options panel if the content overflows the dialog window.

Sorry, nothing have changed.

> Could you check whether resizing the dialog window works? And whether mouse scroll has an effect?

- Resizing works.
- Mouse scroll works only when the scroll bars show up (which is not an unexpected behavior LOL).

I wonder if it is not related to the pre-defined values of the minimum and maximum size of the scroll bar. If the blue bar-cursor of the scroll bar is to high, the minimum size of the window has to be too high for the height of the screen. Setting the height of this blue bar-cursor to a lesser value (If it possible to do that?) should do the trick.

Why do I think that? Because the "Edit" tab needs almost the full height of the screen and when I resize it a little bit less high than the screen I can not see all the content of the tab anymore (even by scrolling).

Actually, even for the tabs with a small height (like the "Tabs" tab), the behavior is the same.

Here are the screenshots.

- "Tabs" tab: resized to just big enough to show everything when scrolling. You can see that I resized to make the horizontal scrollbar only partially visible.

![tabs tab just big enough to show everything when scrolling](https://cloud.githubusercontent.com/assets/20140037/17390438/ac384628-59fc-11e6-880f-c47c0e932e3d.png)

- "Tabs" tab: resized to not enough height to show everything even when scrolling. You can see that the horizontal scrollbar is not visible anymore (even if the content is still scrollable horizontally). The bottom of the vertical scrollbar is not visible neither (you cannot see anymore the two small black arrows).

![tabs tab not enough height to show everything even when scrolling](https://cloud.githubusercontent.com/assets/20140037/17390443/b65c007c-59fc-11e6-95b3-ae8d4ebe3fe1.png)

- "Keys" tab: initial size == full height ==> Not scrollable, doesn't show everything.

![keys tab full height](https://cloud.githubusercontent.com/assets/20140037/17390453/c8b4806e-59fc-11e6-86c4-11692e69c8e9.png)

- "Keys" tab: full height AND resized with less width ==> Scrollable, but doesn't show everything.

![keys tab full height and less width](https://cloud.githubusercontent.com/assets/20140037/17390459/d9a08cba-59fc-11e6-9263-1cf5e116cbd4.png)

- "Edit" tab: before resizing ==> Weird! It is so small!!!

![edit tab before resizing](https://cloud.githubusercontent.com/assets/20140037/17390465/e253b62a-59fc-11e6-9af0-122934bd9544.png)

- "Edit" tab: resized to full height ==> Shows everything.

![edit tab full height](https://cloud.githubusercontent.com/assets/20140037/17390473/ecda8592-59fc-11e6-8ee8-c734e8590ad2.png)

- "Edit" tab: resized to full height, but less width ==> Scrollable. Shows everything.

![edit tab full height but less width](https://cloud.githubusercontent.com/assets/20140037/17390481/f9ff51f8-59fc-11e6-9452-5686c8d49bf0.png)

- "Edit" tab: resized to less height AND less width ==> Scrollable, BUT does NOT show everything anymore.

![edit tab less height and less width](https://cloud.githubusercontent.com/assets/20140037/17390487/06d83b10-59fd-11e6-9a03-79aa42dc5d1a.png)

> Additionally, hotkeys for HTML Editor are also user preferences and are available in the `about:config`. You can set them there if the GUI is unavailable for any reason.

OK. I have used that as a work-around. Actually, even if the GUI is of course more convivial, we don't play often with the customization of the shortcuts, so this way is sufficient for me.

**Comment by danny0838** [comment]
> > Please check `about:config` and see whether `dom.event.contextmenu.enabled` is set to `false`. This will block any attempt to block the context menu from showing up so a right click always shows the context menu.
> 
> `dom.event.contextmenu.enabled` == `true`
> If I have well understood, it is the right value. Isn't it?
> 
> > I'm not sure whether `Control+Click` is also affected by this setting. You can also test it for Control+Click issues.
> 
> Could you be more explicit?

It it suspected that right-click or Ctrl-click (as equivalent to right-click) for the isolate action of DOM Erasor has context menu popup is because your browser has  dom.event.contextmenu.enabled = false, since it's not the case by your report, we'd have to find the true cause in another way.

> `extensions.scrapbook.key.domEraser.narrower3` was set to `Hyphen_Minus`. I have deleted the `_`, setting it to `HyphenMinus`. Now it works again (as it did in 1.13.0b7).

> Actually the key fires a search only if it is not defined as a hot-key, so it should not be a problem. As said above, `-` was not defined because the string `Hyphen_Minus` was not recognized as `-`.

You are right. This is the actual cause. It's a mistake we made during the coding job. 

> >  2. Whether the scrollbar works in the options panel if the content overflows the dialog window.
> 
> Sorry, nothing have changed.
> 
> > Could you check whether resizing the dialog window works? And whether mouse scroll has an effect?
> 
> - Resizing works.
> - Mouse scroll works only when the scroll bars show up (which is not an unexpected behavior LOL).

That bad. It seems that Firefox has some special handling for Macs. We'd need further survay to fix the issue...

**Comment by danny0838** [comment]
We close this issue since the support of command key of Mac (and also Mac style accesskey text) is already done. The issue of overflowing content in the dialog window will be investigated and discussed further in #122.

**Comment by pascallothar** [comment]
@danny0838 
(Because the issue is close, I don't know if I can post here or if I have to post somewhere else. If needed, I can delete it and paste it somewhere else.)

> > I'm not sure whether Control+Click is also affected by this setting. You can also test it for Control+Click issues.
> 
> Could you be more explicit?

Ok, I understand now. I was already unneccessarily "translating" the `Ctrl+Click` in `Cmd+Click` :-D .

In _**DOM Eraser**_ mode, when I use `Ctrl+Click`, it works exactly the same as `RightClick`, i.e. it will _**isolate**_ AND make the **_contextual menu popping up_**.

It isn't the expected behavior, but, as I said before, it is not a big issue.

---

The newly implemented "**_Menu_**" == **_Alt+`C`_** does nothing (i.e. the first shortcut defined in the "Keys" tab). Which "Menu" is it?

Actually, in "HTML Editor" mode, it writes only `©`, which is the default behavior if there is no shortcut defined.

Perhaps is it the implementation of a question I asked?:

> Would it be difficult to implement some Help pop-up in _Scrapbook X HTML Editor_ mode (I would say with shortcut `Cmd+?`) like the one in _Scrapbook X DOM Eraser_ mode?

**Comment by danny0838** [comment]
@pascallothar The menu accesskey refers to the accesskey of the main menu, just like the "F" for file, "E" for edit, and so on. On Windows the default modifier for accesskeys is Alt (Some documentation says it's Alt+Shift. But by my test only Alt works). On Mac it seems to be Control+Option or Control.

Anyway, it seems that we should not simply write "Alt" here anymore. We'd think about another way to present it.

**Comment by pascallothar** [comment]
> The menu accesskey refers to the accesskey of the main menu, just like the "F" for file, "E" for edit, and so on. On Windows the default modifier for accesskeys is Alt (Some documentation says it's Alt+Shift. But by my test only Alt works). On Mac it seems to be Control+Option or Control.
>
> Anyway, it seems that we should not simply write "Alt" here anymore. We'd think about another way to present it.

You learn me something. I have googled and I have found among other stuff [﻿﻿http://superuser.com/questions/303525/what-is-the-shortcut-to-access-the-menubar-in-mac-os-x].

So it seams, indeed, that Windows uses `Alt`.

You say « `Ctrl` or `Ctrl+Option` on Mac », but don't you mistake **_Option key_** (`Alt`) for **_Function modifier key_** (`fn`)? Some settings (like brightness of the screen, sound level, multimedia, ... ) use the function keys (`F1`, `F2`, ... ). By default, you will, for example, access directly (without modifier) to "increase brightness" by hitting the key above the key `2`. To create the event `F2`, you will need to hit this same key with the modifier key `fn` down (`fn+F2` in the documents you have read?). But, in your preferences, you can choose to invert this behavior, i.e. you will hit `F2`to create the event `F2` and `fn+F2` to increase the brightness of the screen. But, if I remember well, isn't it the same on Windows and Linux?

I have just tested it and, indeed, to access the menu bar (the Apple icon on the left of the menu bar to be precise), you need to create the `F2` event, i.e. to hit `fn+F2`by default or `F2` if you made change in your system preferences.

But anyway, it seems that on Mac, the way to go through the menu isn't the same as on Windows and that it is not possible, if I have well understood,  to implement a shortcut for the direct access to the Scrapbook X entry in the menu bar. But is it really necessary, since no other entry of the menu bar is accessed this way on Mac?

**Comment by pascallothar** [comment]
@danny0838 ,
So, my second question is:

Would it be difficult to implement some Help pop-up in _Scrapbook X HTML Editor_ mode (I would say with shortcut `Cmd+?`) like the one in _Scrapbook X DOM Eraser_ mode?

**Comment by pascallothar** [comment]
@danny0838 ,
No answer: Does it mean that you didn't see my question OR that you had too much work and no time OR that **_it would be_** difficult to implement? :-)

Even to make some window (for example, saved as a file in the Scrapbook Directory to let the user fill it or edit it himself) pop up and let to the user the task to create the content of this pop-up?

**Comment by danny0838** [comment]
@pascallothar What content you think should be written in a standalone help page for the HTML editor?

**Comment by pascallothar** [comment]
In "DOM Eraser" mode, when you hit `h`, the help pop-up shows the usage of all the "hotkeys" (not even sure anymore if it is the right word, but anyway). And when you move the mouse or if you hit a second time `h`, the pop-up will disappear.

Here, in "HTML Editor" mode, one could choose a shortcut (Customizable would be better, of course; or something accessible with every keyboard layout and without conflicts with other not customizable shortcuts) that would make the usage window pop up with all the shortcuts used by "HTML Editor". And the window would disappear by moving the mouse. But also, to not use the mouse, by using the shortcut again (or and other one if not easy to implement).

What content? It depends of the implementation. It could be the same type as for "DOM Eraser". But, because the shortcuts are customizable, the pop up can not be "static" and NOT editable at the same time, because it needs to show the actual shortcuts and not the default ones.

If "static", it has to be customizable by the user, even if he has to use a not convivial way to do it (the user doesn't need to do that often, probably only once in its life) (Convivial would be a plus). This way could let more possibilities to the user. For example, for "Format with Template `n`", I would "Format" the string "Format with Template" or a part of the string or even a added word (depending perhaps of the type of formatting). So I could remember easily what template match to what shortcut. Adding an icon would be a plus, but is not a necessity.

If "dynamic" (i.e. the string matching the actual shortcut would be generated dynamically) and NOT editable by the user, one could use what is already shown by the contextual menu, but put together in a single window.
An other possibility would be to create a shortcut that would make the contextual menu pop up immediately with the main sub-menu shown; but is it possible?

"Dynamic" AND editable would let the user add some "custom" tips and tricks, but more important to my eyes, let add some formatted example to the entry "Format with Template `n`".

But, as I said before, it depends of the type of implementation. Editable HTML or XHTML? Or something else? Because, I am not a developer, I have no clue of how to implement it exactly, only some approximative ideas.
